loader panic

## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

loader painc:
`[2021/06/15 11:27:33.117 +08:00] [ERROR] [db.go:177] ["execute statements failed after retry"] [task=test-all] [unit=load] ["worker ID"=10] [queries="[INSERT INTO `dm_meta`.`test-all_loader_checkpoint` (`id`, `filename`, `cp_schema`, `cp_table`, `offset`, `end_pos`) VALUES(?,?,?,?,?,?)]"] [arguments="[[mysql-replica-01 dm_benchmark.sbtest4.0000000020015.sql dm_benchmark sbtest4 0 1048711]]"] [error="[code=10006:class=database:scope=not-set:level=high], Message: execute statement failed: commit, RawCause: Error 1062: Duplicate entry 'mysql-replica-01-dm_benchmark.sbtest4.0000000020015.sql' for key 'uk_id_f'"]
[2021/06/15 11:27:33.119 +08:00] [ERROR] [checkpoint.go:354] ["checkpoint record already exists, skip it."] [task=test-all] [unit=load] [component="remote checkpoint"] [id=mysql-replica-01] [filename=dm_benchmark.sbtest4.0000000020015.sql]
panic: runtime error: index out of range [0] with length 0`

```
goroutine 7918 [running]:
github.com/pingcap/dm/loader.(*RemoteCheckPoint).UpdateOffset(0xc000556200, 0xc001574b95, 0x26, 0xf42f5)
        /Users/ehco/Documents/pingcap/dm/loader/checkpoint.go:406 +0x376
github.com/pingcap/dm/loader.(*Worker).run.func2()
        /Users/ehco/Documents/pingcap/dm/loader/loader.go:202 +0x410
github.com/pingcap/dm/loader.(*Worker).run.func3(0xc0014cae40, 0xc001622020)
        /Users/ehco/Documents/pingcap/dm/loader/loader.go:222 +0x55
created by github.com/pingcap/dm/loader.(*Worker).run
        /Users/ehco/Documents/pingcap/dm/loader/loader.go:220 +0x27e
```



4. Versions of the cluster

    - DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):

        ```
        Release Version: v2.0.0-beta.2-426-g26e4d932
        Git Commit Hash: 26e4d9325935acd61388d2dc31fa2b1dbf9a3047
        Git Branch: lag
        UTC Build Time: 2021-06-09 09:25:27
        Go Version: go version go1.16.3 darwin/amd64
        ```

    - Upstream MySQL/MariaDB server version:

        ```
        5.7.34
        ```

    - Downstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

        ```
        5.0.2
        ```

    - How did you deploy DM: DM-Ansible or manually?

        ```
        manually
        ```


5. current status of DM cluster (execute `query-status` in dmctl)
```
{
    "result": true,
    "msg": "",
    "sources": [
        {
            "result": false,
            "msg": "[code=38029:class=dm-master:scope=internal:level=high], Message: mysql-replica-01 relevant worker-client not found",
            "sourceStatus": {
                "source": "mysql-replica-01",
                "worker": "",
                "result": null,
                "relayStatus": null
            },
            "subTaskStatus": [
            ]
        }
    ]
}
```

6. Operation logs
   - Please upload `dm-worker.log` for every DM-worker instance if possible
   - Please upload `dm-master.log` if possible
   - Other interesting logs
   - Output of dmctl's commands with problems
   
7. Configuration of the cluster and the task
   - `dm-worker.toml` for every DM-worker instance if possible
   - `dm-master.toml` for DM-master if possible
   - task config, like `task.yaml` if possible
   - `inventory.ini` if deployed by DM-Ansible

8. Screenshot/exported-PDF of Grafana dashboard or metrics' graph in Prometheus for DM if possible

--- a/_utils/terror_gen/errors_release.txt
+++ b/_utils/terror_gen/errors_release.txt
@@ -253,7 +253,7 @@ ErrLoadUnitDumpDirNotFound,[code=34014:class=load-unit:scope=internal:level=high
 ErrLoadUnitDuplicateTableFile,[code=34015:class=load-unit:scope=internal:level=high], "Message: invalid table schema file, duplicated item - %s"
 ErrLoadUnitGenBAList,[code=34016:class=load-unit:scope=internal:level=high], "Message: generate block allow list, Workaround: Please check the `block-allow-list` config in task configuration file."
 ErrLoadTaskWorkerNotMatch,[code=34017:class=functional:scope=internal:level=high], "Message: different worker in load stage, previous worker: %s, current worker: %s, Workaround: Please check if the previous worker is online."
-ErrLoadTaskCheckPointNotMatch,[code=34018:class=functional:scope=internal:level=high], "Message: inconsistent checkpoints between loader and target database, Workaround: Please check that you have not forgotten to use dmctl start-task --remove-meta=true."
+ErrLoadTaskCheckPointNotMatch,[code=34018:class=functional:scope=internal:level=high], "Message: inconsistent checkpoints between loader and target database, Workaround: Please check that you have not forgotten to use dmctl start-task --remove-meta=true. if you want to redo the task."
 ErrSyncerUnitPanic,[code=36001:class=sync-unit:scope=internal:level=high], "Message: panic error: %v"
 ErrSyncUnitInvalidTableName,[code=36002:class=sync-unit:scope=internal:level=high], "Message: extract table name for DML error: %s"
 ErrSyncUnitTableNameQuery,[code=36003:class=sync-unit:scope=internal:level=high], "Message: table name parse error: %s"

--- a/_utils/terror_gen/errors_release.txt
+++ b/_utils/terror_gen/errors_release.txt
@@ -253,6 +253,7 @@ ErrLoadUnitDumpDirNotFound,[code=34014:class=load-unit:scope=internal:level=high
 ErrLoadUnitDuplicateTableFile,[code=34015:class=load-unit:scope=internal:level=high], "Message: invalid table schema file, duplicated item - %s"
 ErrLoadUnitGenBAList,[code=34016:class=load-unit:scope=internal:level=high], "Message: generate block allow list, Workaround: Please check the `block-allow-list` config in task configuration file."
 ErrLoadTaskWorkerNotMatch,[code=34017:class=functional:scope=internal:level=high], "Message: different worker in load stage, previous worker: %s, current worker: %s, Workaround: Please check if the previous worker is online."
+ErrLoadTaskCheckPointNotMatch,[code=34018:class=functional:scope=internal:level=high], "Message: inconsistent checkpoints between loader and target database, Workaround: Please check that you have not forgotten to use dmctl start-task --remove-meta=true."
 ErrSyncerUnitPanic,[code=36001:class=sync-unit:scope=internal:level=high], "Message: panic error: %v"
 ErrSyncUnitInvalidTableName,[code=36002:class=sync-unit:scope=internal:level=high], "Message: extract table name for DML error: %s"
 ErrSyncUnitTableNameQuery,[code=36003:class=sync-unit:scope=internal:level=high], "Message: table name parse error: %s"

--- a/_utils/terror_gen/errors_release.txt
+++ b/_utils/terror_gen/errors_release.txt
@@ -253,7 +253,7 @@ ErrLoadUnitDumpDirNotFound,[code=34014:class=load-unit:scope=internal:level=high
 ErrLoadUnitDuplicateTableFile,[code=34015:class=load-unit:scope=internal:level=high], "Message: invalid table schema file, duplicated item - %s"
 ErrLoadUnitGenBAList,[code=34016:class=load-unit:scope=internal:level=high], "Message: generate block allow list, Workaround: Please check the `block-allow-list` config in task configuration file."
 ErrLoadTaskWorkerNotMatch,[code=34017:class=functional:scope=internal:level=high], "Message: different worker in load stage, previous worker: %s, current worker: %s, Workaround: Please check if the previous worker is online."
-ErrLoadTaskCheckPointNotMatch,[code=34018:class=functional:scope=internal:level=high], "Message: inconsistent checkpoints between loader and target database, Workaround: Please check that you have not forgotten to use dmctl start-task --remove-meta=true. if you want to redo the task."
+ErrLoadTaskCheckPointNotMatch,[code=34018:class=functional:scope=internal:level=high], "Message: inconsistent checkpoints between loader and target database, Workaround: If you want to redo the whole task, please check that you have not forgotten to add -remove-meta flag for start-task command."
 ErrSyncerUnitPanic,[code=36001:class=sync-unit:scope=internal:level=high], "Message: panic error: %v"
 ErrSyncUnitInvalidTableName,[code=36002:class=sync-unit:scope=internal:level=high], "Message: extract table name for DML error: %s"
 ErrSyncUnitTableNameQuery,[code=36003:class=sync-unit:scope=internal:level=high], "Message: table name parse error: %s"

--- a/errors.toml
+++ b/errors.toml
@@ -1531,7 +1531,7 @@ tags = ["internal", "high"]
 [error.DM-functional-34018]
 message = "inconsistent checkpoints between loader and target database"
 description = ""
-workaround = "Please check that you have not forgotten to use dmctl start-task --remove-meta=true. if you want to redo the task."
+workaround = "If you want to redo the whole task, please check that you have not forgotten to add -remove-meta flag for start-task command."
 tags = ["internal", "high"]
 
 [error.DM-sync-unit-36001]

--- a/errors.toml
+++ b/errors.toml
@@ -1528,6 +1528,12 @@ description = ""
 workaround = "Please check if the previous worker is online."
 tags = ["internal", "high"]
 
+[error.DM-functional-34018]
+message = "inconsistent checkpoints between loader and target database"
+description = ""
+workaround = "Please check that you have not forgotten to use dmctl start-task --remove-meta=true."
+tags = ["internal", "high"]
+
 [error.DM-sync-unit-36001]
 message = "panic error: %v"
 description = ""

--- a/errors.toml
+++ b/errors.toml
@@ -1531,7 +1531,7 @@ tags = ["internal", "high"]
 [error.DM-functional-34018]
 message = "inconsistent checkpoints between loader and target database"
 description = ""
-workaround = "Please check that you have not forgotten to use dmctl start-task --remove-meta=true."
+workaround = "Please check that you have not forgotten to use dmctl start-task --remove-meta=true. if you want to redo the task."
 tags = ["internal", "high"]
 
 [error.DM-sync-unit-36001]

--- a/loader/checkpoint.go
+++ b/loader/checkpoint.go
@@ -400,7 +400,7 @@ func (cp *RemoteCheckPoint) UpdateOffset(filename string, offset int64) error {
 	defer cp.restoringFiles.Unlock()
 	db, table, err := getDBAndTableFromFilename(filename)
 	if err != nil {
-		return terror.ErrLoadTaskCheckPointNotMatch.Generate()
+		return terror.Annotatef(terror.ErrLoadTaskCheckPointNotMatch.Generate(), "wrong filename=%s", filename)
 	}
 
 	if _, ok := cp.restoringFiles.pos[db]; ok {
@@ -411,7 +411,7 @@ func (cp *RemoteCheckPoint) UpdateOffset(filename string, offset int64) error {
 			}
 		}
 	}
-	return terror.ErrLoadTaskCheckPointNotMatch.Generate()
+	return terror.Annotatef(terror.ErrLoadTaskCheckPointNotMatch.Generate(), "db=%s table=%s not in checkpoint", db, filename)
 }
 
 // Clear implements CheckPoint.Clear.

--- a/loader/checkpoint.go
+++ b/loader/checkpoint.go
@@ -73,7 +73,7 @@ type CheckPoint interface {
 
 	// UpdateOffset keeps `cp.restoringFiles` in memory same with checkpoint in DB,
 	// should be called after update checkpoint in DB
-	UpdateOffset(filename string, offset int64)
+	UpdateOffset(filename string, offset int64) error
 
 	// AllFinished returns `true` when all restoring job are finished
 	AllFinished() bool
@@ -395,15 +395,23 @@ func (cp *RemoteCheckPoint) GenSQL(filename string, offset int64) string {
 }
 
 // UpdateOffset implements CheckPoint.UpdateOffset.
-func (cp *RemoteCheckPoint) UpdateOffset(filename string, offset int64) {
+func (cp *RemoteCheckPoint) UpdateOffset(filename string, offset int64) error {
 	cp.restoringFiles.Lock()
 	defer cp.restoringFiles.Unlock()
 	db, table, err := getDBAndTableFromFilename(filename)
 	if err != nil {
-		cp.logger.Error("error in checkpoint UpdateOffset", zap.Error(err))
-		return
+		return terror.ErrLoadTaskCheckPointNotMatch.Generate()
 	}
-	cp.restoringFiles.pos[db][table][filename][0] = offset
+
+	if _, ok := cp.restoringFiles.pos[db]; ok {
+		if _, ok := cp.restoringFiles.pos[db][table]; ok {
+			if _, ok := cp.restoringFiles.pos[db][table][filename]; ok {
+				cp.restoringFiles.pos[db][table][filename][0] = offset
+				return nil
+			}
+		}
+	}
+	return terror.ErrLoadTaskCheckPointNotMatch.Generate()
 }
 
 // Clear implements CheckPoint.Clear.

--- a/loader/checkpoint.go
+++ b/loader/checkpoint.go
@@ -400,7 +400,7 @@ func (cp *RemoteCheckPoint) UpdateOffset(filename string, offset int64) error {
 	defer cp.restoringFiles.Unlock()
 	db, table, err := getDBAndTableFromFilename(filename)
 	if err != nil {
-		return terror.Annotatef(terror.ErrLoadTaskCheckPointNotMatch.Generate(), "wrong filename=%s", filename)
+		return terror.Annotatef(terror.ErrLoadTaskCheckPointNotMatch.Generate(err), "wrong filename=%s", filename)
 	}
 
 	if _, ok := cp.restoringFiles.pos[db]; ok {

--- a/loader/checkpoint.go
+++ b/loader/checkpoint.go
@@ -411,7 +411,7 @@ func (cp *RemoteCheckPoint) UpdateOffset(filename string, offset int64) error {
 			}
 		}
 	}
-	return terror.Annotatef(terror.ErrLoadTaskCheckPointNotMatch.Generate(), "db=%s table=%s not in checkpoint", db, filename)
+	return terror.ErrLoadTaskCheckPointNotMatch.Generatef("db=%s table=%s not in checkpoint", db, filename)
 }
 
 // Clear implements CheckPoint.Clear.

--- a/loader/loader.go
+++ b/loader/loader.go
@@ -217,9 +217,7 @@ func (w *Worker) run(ctx context.Context, fileJobQueue chan *fileJob, runFatalCh
 				job.file = "notafile" + job.file
 			})
 			if err := w.loader.checkPoint.UpdateOffset(job.file, job.offset); err != nil {
-				if !utils.IsContextCanceledError(err) {
-					runFatalChan <- unit.NewProcessError(err)
-				}
+				runFatalChan <- unit.NewProcessError(err)
 				hasError = true
 				continue
 			}

--- a/loader/loader.go
+++ b/loader/loader.go
@@ -212,7 +212,17 @@ func (w *Worker) run(ctx context.Context, fileJobQueue chan *fileJob, runFatalCh
 				})
 				continue
 			}
-			w.loader.checkPoint.UpdateOffset(job.file, job.offset)
+
+			failpoint.Inject("loaderCPUpdateOffsetError", func(_ failpoint.Value) {
+				job.file = "notafile" + job.file
+			})
+			if err := w.loader.checkPoint.UpdateOffset(job.file, job.offset); err != nil {
+				if !utils.IsContextCanceledError(err) {
+					runFatalChan <- unit.NewProcessError(err)
+				}
+				hasError = true
+				continue
+			}
 			w.loader.finishedDataSize.Add(job.offset - job.lastOffset)
 		}
 	}

--- a/pkg/terror/error_list.go
+++ b/pkg/terror/error_list.go
@@ -944,7 +944,7 @@ var (
 	ErrLoadUnitDuplicateTableFile  = New(codeLoadUnitDuplicateTableFile, ClassLoadUnit, ScopeInternal, LevelHigh, "invalid table schema file, duplicated item - %s", "")
 	ErrLoadUnitGenBAList           = New(codeLoadUnitGenBAList, ClassLoadUnit, ScopeInternal, LevelHigh, "generate block allow list", "Please check the `block-allow-list` config in task configuration file.")
 	ErrLoadTaskWorkerNotMatch      = New(codeLoadTaskWorkerNotMatch, ClassFunctional, ScopeInternal, LevelHigh, "different worker in load stage, previous worker: %s, current worker: %s", "Please check if the previous worker is online.")
-	ErrLoadTaskCheckPointNotMatch  = New(codeLoadCheckPointNotMatch, ClassFunctional, ScopeInternal, LevelHigh, "inconsistent checkpoints between loader and target database", "Please check that you have not forgotten to use dmctl start-task --remove-meta=true.")
+	ErrLoadTaskCheckPointNotMatch  = New(codeLoadCheckPointNotMatch, ClassFunctional, ScopeInternal, LevelHigh, "inconsistent checkpoints between loader and target database", "Please check that you have not forgotten to use dmctl start-task --remove-meta=true. if you want to redo the task.")
 
 	// Sync unit error.
 	ErrSyncerUnitPanic                   = New(codeSyncerUnitPanic, ClassSyncUnit, ScopeInternal, LevelHigh, "panic error: %v", "")

--- a/pkg/terror/error_list.go
+++ b/pkg/terror/error_list.go
@@ -944,7 +944,7 @@ var (
 	ErrLoadUnitDuplicateTableFile  = New(codeLoadUnitDuplicateTableFile, ClassLoadUnit, ScopeInternal, LevelHigh, "invalid table schema file, duplicated item - %s", "")
 	ErrLoadUnitGenBAList           = New(codeLoadUnitGenBAList, ClassLoadUnit, ScopeInternal, LevelHigh, "generate block allow list", "Please check the `block-allow-list` config in task configuration file.")
 	ErrLoadTaskWorkerNotMatch      = New(codeLoadTaskWorkerNotMatch, ClassFunctional, ScopeInternal, LevelHigh, "different worker in load stage, previous worker: %s, current worker: %s", "Please check if the previous worker is online.")
-	ErrLoadTaskCheckPointNotMatch  = New(codeLoadCheckPointNotMatch, ClassFunctional, ScopeInternal, LevelHigh, "inconsistent checkpoints between loader and target database", "Please check that you have not forgotten to use dmctl start-task --remove-meta=true. if you want to redo the task.")
+	ErrLoadTaskCheckPointNotMatch  = New(codeLoadCheckPointNotMatch, ClassFunctional, ScopeInternal, LevelHigh, "inconsistent checkpoints between loader and target database", "If you want to redo the whole task, please check that you have not forgotten to add -remove-meta flag for start-task command.")
 
 	// Sync unit error.
 	ErrSyncerUnitPanic                   = New(codeSyncerUnitPanic, ClassSyncUnit, ScopeInternal, LevelHigh, "panic error: %v", "")

--- a/pkg/terror/error_list.go
+++ b/pkg/terror/error_list.go
@@ -344,6 +344,7 @@ const (
 	codeLoadUnitDuplicateTableFile
 	codeLoadUnitGenBAList
 	codeLoadTaskWorkerNotMatch
+	codeLoadCheckPointNotMatch
 )
 
 // Sync unit error code.
@@ -943,6 +944,7 @@ var (
 	ErrLoadUnitDuplicateTableFile  = New(codeLoadUnitDuplicateTableFile, ClassLoadUnit, ScopeInternal, LevelHigh, "invalid table schema file, duplicated item - %s", "")
 	ErrLoadUnitGenBAList           = New(codeLoadUnitGenBAList, ClassLoadUnit, ScopeInternal, LevelHigh, "generate block allow list", "Please check the `block-allow-list` config in task configuration file.")
 	ErrLoadTaskWorkerNotMatch      = New(codeLoadTaskWorkerNotMatch, ClassFunctional, ScopeInternal, LevelHigh, "different worker in load stage, previous worker: %s, current worker: %s", "Please check if the previous worker is online.")
+	ErrLoadTaskCheckPointNotMatch  = New(codeLoadCheckPointNotMatch, ClassFunctional, ScopeInternal, LevelHigh, "inconsistent checkpoints between loader and target database", "Please check that you have not forgotten to use dmctl start-task --remove-meta=true.")
 
 	// Sync unit error.
 	ErrSyncerUnitPanic                   = New(codeSyncerUnitPanic, ClassSyncUnit, ScopeInternal, LevelHigh, "panic error: %v", "")

--- a/tests/load_interrupt/run.sh
+++ b/tests/load_interrupt/run.sh
@@ -33,12 +33,43 @@ function check_row_count() {
 	[ "$estimate" == "$row_count" ]
 }
 
-function run() {
-	THRESHOLD=1024
+function test_save_checkpoint_faild() {
 	prepare_datafile
-
 	run_sql_file $WORK_DIR/db1.prepare.sql $MYSQL_HOST1 $MYSQL_PORT1 $MYSQL_PASSWORD1
 
+	export GO_FAILPOINTS="github.com/pingcap/dm/loader/loaderCPUpdateOffsetError=return(1)"
+
+	run_dm_master $WORK_DIR/master $MASTER_PORT $cur/conf/dm-master.toml
+	check_rpc_alive $cur/../bin/check_master_online 127.0.0.1:$MASTER_PORT
+	run_dm_worker $WORK_DIR/worker1 $WORKER1_PORT $cur/conf/dm-worker1.toml
+	check_rpc_alive $cur/../bin/check_worker_online 127.0.0.1:$WORKER1_PORT
+	# operate mysql config to worker
+	cp $cur/conf/source1.yaml $WORK_DIR/source1.yaml
+	sed -i "/relay-binlog-name/i\relay-dir: $WORK_DIR/worker1/relay_log" $WORK_DIR/source1.yaml
+	dmctl_operate_source create $WORK_DIR/source1.yaml $SOURCE_ID1
+
+	run_dm_ctl $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+		"start-task $cur/conf/dm-task.yaml"
+
+	# load task should Paused because job file is not right
+	run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+		"query-status test" \
+		"Paused" 1
+
+	# check dump files are generated before worker down
+	ls $WORK_DIR/worker1/dumped_data.test
+
+	echo "test_save_checkpoint_faild SUCCESS!"
+	cleanup_data load_interrupt
+	cleanup_process $*
+}
+
+function run() {
+	test_save_checkpoint_faild
+
+	prepare_datafile
+	run_sql_file $WORK_DIR/db1.prepare.sql $MYSQL_HOST1 $MYSQL_PORT1 $MYSQL_PASSWORD1
+	THRESHOLD=1024
 	export GO_FAILPOINTS="github.com/pingcap/dm/loader/LoadExceedOffsetExit=return($THRESHOLD)"
 
 	run_dm_master $WORK_DIR/master $MASTER_PORT $cur/conf/dm-master.toml

--- a/tests/load_interrupt/run.sh
+++ b/tests/load_interrupt/run.sh
@@ -37,7 +37,7 @@ function test_save_checkpoint_faild() {
 	prepare_datafile
 	run_sql_file $WORK_DIR/db1.prepare.sql $MYSQL_HOST1 $MYSQL_PORT1 $MYSQL_PASSWORD1
 
-	export GO_FAILPOINTS="github.com/pingcap/dm/loader/loaderCPUpdateOffsetError=return(1)"
+	export GO_FAILPOINTS="github.com/pingcap/dm/loader/loaderCPUpdateOffsetError=return()"
 
 	run_dm_master $WORK_DIR/master $MASTER_PORT $cur/conf/dm-master.toml
 	check_rpc_alive $cur/../bin/check_master_online 127.0.0.1:$MASTER_PORT

